RandHound 341

Daeinar opened the pull request:
Addresses  #341 and #321.

ineiti commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+
+// Setup initializes the proof by randomly selecting a commitment v,
+// determining the challenge c = H(xG,xH,vG,vH), and the response r = v - cx.
+func (p *Proof) Setup(scalar ...abstract.Scalar) ([]abstract.Point, []abstract.Point, []ProofCore, error) {
```
Would it make sense to put the `[]abstract.Point` together in a structure?

Daeinar replied:
I am not sure. I tried to keep it as simple as possible and I am not sure if you get any benefit of having just `[]abstract.Point` in a structure. You would probably have to put all the return values into a structure. I guess that's what you mean? Iirc I tried that a while ago and it complicated quite a few things in the other code that uses PVSS. I'll think about it again.

nikkolasg replied:
> You would probably have to put all the return values into a structure.

Yes, that'd be nice if possible. If you remember which part posed problems, don't hesitate to drop it to us to see if something's possible.

ineiti commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+		return nil, nil, errors.New("Received unexpected number of points")
+	}
+
+	var good, bad []int
```
The good, the bad, and the ugly.

ineiti commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+	}
+
+	var good, bad []int
+	for i := 0; i < len(p.base); i++ {
```
`for i, core := range p.core`
and replace `p.core[i]` with `core`

Daeinar replied:
You probably mean `for i := range p.base {...}`?

ineiti commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+	n := len(X)
+
+	// Create secret sharing polynomial
+	p := new(poly.PriPoly).Pick(pv.suite, pv.t, secret, random.Stream)
```
Possibility of renaming `p`, `s` and `P` to `poly`, `shares`, `commit`, so you don't need comments.

ineiti commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+	share := make([]abstract.Scalar, n)
+	H := make([]abstract.Point, n)
+	idx := make([]int, n)
+	for i := 0; i < n; i++ {
```
`for i := range idx`

ineiti commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+
+	n := len(polyBin)
+	sH := make([]abstract.Point, n)
+	for i := 0; i < n; i++ {
```
`for i := range sH`

ineiti commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+	pp := new(poly.PubPoly).InitNull(pv.suite, pv.t, pv.suite.Point().Base())
+	ps := new(poly.PubShares).Split(pp, n) // XXX: ackward way to init shares
+
+	for i := 0; i < len(S); i++ {
```
```
for i, s := range S{
  ps.SetShare(pos[i], s)
}
```

ineiti commented on protocols/randhound/proof_test.go:
```diff
@@ -0,0 +1,190 @@
+	g := []abstract.Point{g1, g2}
+	h := []abstract.Point{h1, h2}
+	p, err := randhound.NewProof(suite, g, h, nil)
+	if err != nil {
```
Possibility of using
`log.ErrFatal(err)`
which `fatal`s if `err != nil`
or use
`require.Nil(t, err)`

Liamsi replied:
@Daeinar If you use `log.ErrFatal(err)` as @ineiti suggested, you don't need to check `err != nil` (because it's done in [`ErrFatal`](https://github.com/dedis/cothority/blob/08796f10dfa5b418c0deae99bcb12fac08f0d846/log/ui.go#L83)). Personally, I don't care much about the code style in the tests...

ineiti commented on protocols/randhound/proof_test.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	if len(bad) != 0 {
+		t.Fatal(fmt.Errorf("Some proofs failed: %v", bad))
```
`log.Fatalf("Some proofs failed: %v", bad)`

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
+		rh.group[i] = g
+	}
+
+	// Comptue session id
```
s/Comptue/Compute/

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
+
+		// Sign I1 and store signature in i1.Sig
+		if err := signSchnorr(rh.Suite(), rh.Private(), i1); err != nil {
+			return err
```
mutex will be held here. Add `Unlock`

Liamsi replied:
I think @ineiti meant the mutex won't be released in the error case. Not sure if this is problematic but to be sure you could add a `Unlock` in this branch:
```
if err := signSchnorr(rh.Suite(), rh.Private(), i1); err != nil {
          rh.mutex.Unlock()
          return err
}
```

Daeinar replied:
Ah of course, that makes sense. Thanks for clarifying!

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
+				}
+			}
+
+			//log.Lvlf1("Env vs Dec (Pre-Sync):  %v %v %v %v %v %v %v", encPos, decPos, len(X), len(encShare), len(encProof), len(decShare), len(decProof))
```
Is this still used?

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
+				X = X[:l]
+			}
+
+			//log.Lvlf1("Env vs Dec (Post-Sync): %v %v %v %v %v %v %v", encPos, decPos, len(X), len(encShare), len(encProof), len(decShare), len(decProof))
```
Still used? Can also write `log.Lvlf3` or so.

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
+			_ = goodEnc
+			_ = badEnc
+
+			//log.Lvlf1("Enc: %v %v %v %v %v %v", goodEnc, badEnc, len(X), len(encShare), len(decShare), len(decProof))
```
Still used?

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
+				decShare = append(decShare[:k], decShare[k+1:]...)
+			}
+
+			//log.Lvlf1("Dec: %v %v", goodDec, badDec)
```
Is this still used?

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
+				return err
+			}
+			rnd = rh.Suite().Point().Add(rnd, ps)
+			//log.Lvlf1("Transcript: %v %v", src, ps)
```
Is this still used?

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
-		return nil, nil, err
+
+	// Record valid encrypted shares per secret/server
+	for i := 0; i < len(good); i++ {
```
`for i, g := range good`
and replace `good[i]` with `g`.

ineiti commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,1006 @@
+			rh.chosenSecret[i] = secret
+		}
+
+		//log.Lvlf1("Grouping: %v", rh.group)
```
Still used? Possible to use `log.Lvlf3`?

Daeinar commented:
@ineiti: Addressed most of your comments. I did not change the format of the return values in `Setup(...)`. I guess we can do that when we resolve #624.

nikkolasg commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,995 @@
+	for i := range m {
+		j := int(random.Uint64(prng) % uint64(i+1))
+		m[i] = m[j]
+		m[j] = i + 1
```
I don't know where this comes from but it deviates a bit from the "Fisher Yates " shuffle then derived by Durstenfeld for CS ( see [link](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_.22inside-out.22_algorithm) ).
The part that seems missing is the `if i != j { m[i] = m[j] }`. Of course I haven't checked out the recursion proof of the wikipedia algo so I would have to spend more time to check if this impl. is correct. 
How did you came up with this implementation ?

Daeinar replied:
It's basically an adaptation of Golang's `rand.Perm(...)` function, see here: https://golang.org/src/math/rand/rand.go?s=5071:5103#L164

Daeinar replied:
Ah, yes that makes sense!

nikkolasg commented:
Ok this is a huge PR and I'm not sure to be familiar enough with all the crypto inside to be a really good reviewer for this one. I'd need to read the paper on JVSS and the one you're writing to get a full grasp. I'm planning on this but now now so on the surface, to me the code looks good :)

On a more structural level, I see you coded PVSS yourself in the same package. Would it make sense now / later to decouple that from the randhound code and to integrate that in the crypto library ? I think it would be a real good target for the crypto library, the same way we have jvss & cosi integrated in the crypto library. Of course these were introduced after we coded the protocol so it does not have to be now. What do you think ?

Daeinar commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,995 @@
+	hs := rh.Suite().Hash().Size()
+	rand := make([]byte, hs)
+	random.Stream.XORKeyStream(rand, rand)
+	rh.cliRand = rand
```
Ah thanks I didn't know about this function in dedis/crypto, I will of course use that instead.

Daeinar commented on protocols/randhound/randhound.go:
```diff
@@ -1,251 +1,995 @@
+				// Check availability of corresponding R2 messages, skip if not there
+				target := r1.EncShare[j].Target
+				if _, ok := t.R2s[target]; !ok {
+					continue OUTER
```
Note that the r1 message changes the next time you go to `OUTER`, see https://github.com/dedis/cothority/blob/randhound_341/protocols/randhound/randhound.go#L366.

Daeinar replied:
@nikkolasg: Okay leaving out the `OUTER` also works. ;)

Daeinar commented:
@nikkolasg: Fully agree on the PVSS part. I've already added an issue on that topic a couple of days ago, see #624. I am not sure how urgent that is though and we maybe first want to focus on other things?

Liamsi commented on crypto/hash.go:
```diff
@@ -118,22 +118,21 @@ func HashArgsSuite(suite abstract.Suite, args ...interface{}) ([]byte, error) {
-// a slice of BinaryMarshalers.
-func convertToBinaryMarshaler(args ...interface{}) ([]encoding.BinaryMarshaler, error) {
+// a slice of BinaryMarshalers
+func ConvertToBinaryMarshaler(args ...interface{}) ([]encoding.BinaryMarshaler, error) {
```
Will this be used externally (would make sense). If not, I think this should remain un-exported.

Daeinar replied:
We will move all of the stuff from `crypto/hash.go` to the crypto library at some point. So I this can be addressed later.

Liamsi commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+
+// Verify validates the proof against the given input by checking that
+// vG == rG + c(xG) and vH == rH + c(xH).
+func (p *Proof) Verify(xG []abstract.Point, xH []abstract.Point) ([]int, []int, error) {
```
I haven't read the paper of the algo you are implementing here. But this `Verify` function looks like something an external dev/PhD student would use. It would be great if was clearer for the caller when this `Proof` is valid or not. From looking into the implementation and the naming of the local vars (`good`, `bad`) it looks like the proof is valid iff the second returned array is empty? (OK, looking into the tests clarifies this). But you could either use named vars or if you prefer comments, you could add a short comment, when this `Proof` is a valid one or not.

Daeinar replied:
Yes good point, I'll fix the comments! Since we are doing threshold crypto here, it is allowed that some proofs fail, which is also the reason for the `good` and `bad` return values.

Liamsi commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+}
+
+// Verify checks that log_H(sH) == log_X(sX) using the given proof.
+func (pv *PVSS) Verify(H abstract.Point, X []abstract.Point, sH []abstract.Point, sX []abstract.Point, core []ProofCore) ([]int, []int, error) {
```
Same comment as https://github.com/dedis/cothority/pull/519/files#r82609495

Liamsi commented:
Besides a few comments this PR looks good to me. Feel free to decide which are relevant to you (and which not).

Daeinar commented on protocols/randhound/struct.go:
```diff
@@ -0,0 +1,137 @@
+	// Group information
+	server              [][]*sda.TreeNode  // Grouped servers
+	group               [][]int            // Grouped server indices
+	threshold           []int              // Group thresholds
```
I am usually sticking to the singular, so that when you iterate over the variable you have `threshold[i]` instead of `thresholds[i]` which is more intuitive to me. In the above example the same holds for `server[i][j]` etc. I should probably check if there are some Golang coding conventions on that.

Liamsi replied:
I think the most idiomatic golang way would be to iterate using `range` instead. Like here: https://play.golang.org/p/HtZlb8Ctav
But as I said, this was just nitpicking ;-)

Daeinar replied:
Hmm right. I just get used to that. ;)

ineiti replied:
+1 for `range` ;)

Liamsi commented on protocols/randhound/proof.go:
```diff
@@ -0,0 +1,298 @@
+		p.core[i] = ProofCore{c, r, vG, vH}
+	}
+
+	return xG, xH, p.core, nil
```
> Would it be better coding style to have a "getter"?

I think this depends on your preferences. As `p.core` is an array, the returned object (in any case) can be manipulated. So you could as well just make `core` a public/exported field:
See here: https://play.golang.org/p/pkZHSVzSgk

Liamsi replied:
And the reason I suggested that isn't about the getter, it's just to make the code more readable by reducing the number of returned objects (as the caller could have access to `core` by naming it `Core` without returning it and the provided "safety" would be the same ). Mainly because the others complained about the number of args. But if you don't like it, just keep it like this. It's just a matter of style (and I'm sure many people could have vim vs emacs like discussion about it ;-)
